### DISCUSSION TITLE
What is the best way to add custom columns to Claims and Credentials

### DISCUSSION THREAD

**Discussion by Muhammad-Altabba** [decentralized-identity/veramo#647] [Q&A] [answered]
Hello,
We are using the Web DID method. And we need to have custom columns to save for example the relevant id of the claim that exists in our centralized system. And another column to indicate if the credential is supposed to be publicly available through an endpoint that will list all public credentials (like the official website of the company for example).
What do you suggest for us to do? And are those columns are this the best to be added to Claims or to Credentials? 
Thanks in advance,

**Comment by mirceanis** [comment, marked as answer]
Hi @Muhammad-Altabba,
Are these custom claims embedded in the credentials? If they are, you don't need any modification to the database layout. They can be matched using the existing @veramo/data-store API.

Also, the DID method that you use should not matter for this.

**Reply by Muhammad-Altabba** [reply]
Many Thanks @mirceanis ,
Yes, I think those could be embedded in the credentials.
However, could you please give me more hints about the @veramo/data-store API that I can use? 
I checked for example `createVerifiableCredential` but it does not save the provided custom attributes for the `credential` object. And I could not find what I should call to add a custom column. It would be great if you give me a link to a code snippet or documentation...

Or do you direct me to save this custom attribute inside the `credentialSubject` of the `credential`? In that case, how to query for example based on a specific value of the custom column?

**Reply by Muhammad-Altabba** [reply]
Hello @mirceanis 
If you could please give us some of your valuable hints 😊 
Many thanks,

**Reply by mirceanis** [reply]
Hi @Muhammad-Altabba,
You can try to embed the extra attributes as claims in `credentialSubject` when you create credentials:

```typescript
const credential = await agent.createVerifiableCredential({
  credential: {
    issuer: { id: issuer.did },
    type: ['VerifiableCredential'],
    credentialSubject: {
      id: subject.did,
      hello: "world" // add custom properties as claims here
    },
  },
  proofFormat: 'jwt',
  save: true,
})
```

Then, when you need to get the credentials that have these custom claims, you can try something like this:
```ts
const filteredCredentials = await agent.dataStoreORMGetVerifiableCredentialsByClaims({
  where: [
    { column: 'type', value: ['hello'] },
    { column: 'value', value: ['world'] },
  ],
})
```

I hope this helps

**Comment by martijndebruijn** [comment]
What if the additional attributes are not part of the Credential (o not part of the proof)? I need to store some additional attrbutes. It would be helpful if I could store them with the Credential.

**Reply by mirceanis** [reply]
Then I suppose a `metadata` property for a credential would be helpful.

We don't have this now, but it can technically be added.

Can you share an example of how you would use this?

**Comment by martijndebruijn** [comment]
A meta data property would help a lot. Currently we store some internal
state (status, exteralKey) which is not part of the VC schema. We use the
additional internal state in the Android/iOS app to provide some
additional functionality.




Op zo 21 nov. 2021 om 15:31 schreef Mircea Nistor ***@***.***
